### PR TITLE
[Storage] Matching `_shared` folder contents across all packages

### PR DESCRIPTION
**Pull request by vincenttran-msft**
This PR aims to unify all the contents of the `_shared` folders for the blob, datalake, queue, and file-share packages.

In general, my process was to use blob's `_shared` folder as the "source of truth" for most things as it is where most new changes will be added (and then not uniformly added to the other `_shared` folders)

Things to callout when unifying:

- Biggest change came from: Support for Oauth import/export managed disks [here](https://github.com/Azure/azure-sdk-for-python/pull/22984). I am assuming this will go everywhere eventually so most of the work was getting the shared folders to all contain these changes (please correct me if this won't go everywhere, then I will have to make all the packages look similar to each other excluding this change 😢)
- `shared_access_signature.py` will be different for file-share and queue since they do **not** support encryption scope
- `parser.py` is mostly barren except for in blob -- I opted to copy blob's `parser.py` to every other file (which is likely overkill) just because I anticipate most new features in blob that require these parser changes will eventually trickle down to other packages
- In `response_handlers.py` metadata parsing went from `response.headers.items` to `response.http_reponse.headers.items`. Not sure how this will play out because I'm assuming if it works in its current state, will the same information be contained in `response.http_response` vs `response.headers`? Pending pipeline run for the answer 😄 
- In `policies.py` specifically the `on_request` method, the `queue` package is the only file with this workaround (and thus will differ from other files):
```
# Hack to fix generated code adding '/messages' after SAS parameters
        includes_messages = request.http_request.url.endswith('/messages')
        if includes_messages:
            request.http_request.url = request.http_request.url[:-(len('/messages'))]
            request.http_request.url = urljoin(request.http_request.url, 'messages')
```

Therefore, the only differences in the `_shared` folders now is:
- Queue generated code hack
- `shared_access_signature.py` differences for packages that don't support encryption scope (file-share & queue)
- Docstring differences that callout the package name

### CODE REVIEW

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/response_handlers.py [hunk 1]
```diff
@@ -67,7 +67,7 @@ def normalize_headers(headers):
 
 
 def deserialize_metadata(response, obj, headers):  # pylint: disable=unused-argument
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/response_handlers.py:69]: Check on usage in datalake

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/shared_access_signature.py [hunk 2]
```diff
@@ -39,6 +39,8 @@ class QueryStringConstants(object):
     SIGNED_KEY_EXPIRY = 'ske'
     SIGNED_KEY_SERVICE = 'sks'
     SIGNED_KEY_VERSION = 'skv'
+
+    # for blob only
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/shared_access_signature.py:43]: Fix these

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/uploads.py [hunk 3]
```diff
@@ -598,10 +598,11 @@ def read(self, size):
                     chunk = chunk.encode(self.encoding)
                 data += chunk
                 count += len(chunk)
+        # This means count < size and what's leftover will be returned in this call.
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/uploads.py:601]: Can you take a look at this change in Blob and see what it was about?

File: sdk/storage/azure-storage-queue/azure/storage/queue/_shared/authentication.py [hunk 4]
```diff
@@ -28,17 +30,22 @@
...
 logger = logging.getLogger(__name__)
 
 
-
 # wraps a given exception with the desired exception type
 def _wrap_exception(ex, desired_type):
     msg = ""
     if ex.args:
         msg = ex.args[0]
-    return desired_type(msg)
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-queue/azure/storage/queue/_shared/authentication.py:41]: Should be the old change. Apply to other packages.

File: sdk/storage/azure-storage-queue/azure/storage/queue/_shared/base_client.py [hunk 5]
```diff
@@ -80,9 +82,9 @@ def __init__(
         self.scheme = parsed_url.scheme
 
         if service not in ["blob", "queue", "file-share", "dfs"]:
-            raise ValueError(f"Invalid service: {service}")
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-queue/azure/storage/queue/_shared/base_client.py:83]: `f-string` is the correct thing not `.format`.

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/uploads.py (hunk 3 above)
Comment (vincenttran-msft) [reply sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/uploads.py:601]: https://github.com/Azure/azure-sdk-for-python/pull/17551

This is what it was from (bugfix for a infinite looping generator bug) so seems like it is a good thing to duplicate across other packages?

File: sdk/storage/azure-storage-queue/azure/storage/queue/_shared/response_handlers.py [hunk 6]
```diff
@@ -83,16 +83,18 @@ def return_context_and_deserialized(response, deserialized, response_headers):
...
 
 
+def return_raw_deserialized(response, *_):
+    return response.http_response.location_mode, response.context[ContentDecodePolicy.CONTEXT_NAME]
+
+
 def process_storage_error(storage_error):   # pylint:disable=too-many-statements
     raise_error = HttpResponseError
     serialized = False
-    if not storage_error.response or storage_error.response.status_code in [200, 204]:
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-queue/azure/storage/queue/_shared/response_handlers.py:89]: Chatted offline about looking into this -- 
Going to revert this to how it was previously as it's been like that since even before this PR that moved it: 
https://github.com/tasherif-msft/azure-sdk-for-python/commit/397c70b4e4e384bde14d3777fe81eef5e64f4d63

File: sdk/storage/azure-storage-queue/azure/storage/queue/_shared/request_handlers.py [hunk 7]
```diff
@@ -71,7 +71,11 @@ def get_length(data):
             pass
         else:
             try:
-                return fstat(fileno).st_size
+                mode = fstat(fileno).st_mode
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-queue/azure/storage/queue/_shared/request_handlers.py:74]: This was a bugfix for blob (upload_blob): https://github.com/Azure/azure-sdk-for-python/pull/19725 should be fine to propogate elsewhere?

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/response_handlers.py (hunk 1 above)
Comment (vincenttran-msft) [reply sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/response_handlers.py:69]: So I'm not exactly sure when this `deserialize_metadata` [method](https://github.com/Azure/azure-sdk-for-python/blob/e21e984c0752cd05ab51d66940afb1c0b9b6abbf/sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/response_handlers.py#L69) would be called as the only instance I see of metadata serialization is in `get_file_properties()` or `get_directory_properties()`. 


get_file_properties: `return self._get_path_properties(cls=deserialize_file_properties, **kwargs)`

Following this debugger trail will take you to `_deserialize.py`'s `deserialize_file_properties` method, which in turn will call its `deserialize_metadata` (and **NOT** the one defined in `response_handlers.py` found [here](https://github.com/Azure/azure-sdk-for-python/blob/e21e984c0752cd05ab51d66940afb1c0b9b6abbf/sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_deserialize.py#L110)

The same is true for directories (will call into `_deserialize.py`'s methods and **NOT** `response_handlers.py`'s methods.

Furthermore, looking through datalake's imports from `from ._shared.response_handlers`, shows that none of them ever import `deserialize_metadata` explicitly (and no methods that _are_ imported from this file call `deserialize_metadata` as a subroutine. 

**tl;dr** - This is **unused**, so we can change it to keep _shared consistent, and if it does start being used it will error out and so all of them will be revisited to fix functionality + be the same across all packages

Comment (vincenttran-msft) [reply sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/response_handlers.py:69]: In contrast, blob's `get_properties` call does call into `response_handlers.py`'s `deserialize_metadata` method, and this is the response object (hence why it needs .http_response):
image

This is an example of datalake's response object:
image

So blob's is a `PipelineResponse` which contains a `RequestsTransportResponse` while datalake is just the `RequestsTransportResponse`

File: sdk/storage/azure-storage-queue/azure/storage/queue/_shared/uploads_async.py [hunk 8]
```diff
@@ -296,7 +296,7 @@ async def _upload_chunk(self, chunk_offset, chunk_data):
 
     async def _upload_substream_block(self, index, block_stream):
         try:
-            block_id = f'BlockId{index/self.chunk_size:05d}'
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-queue/azure/storage/queue/_shared/uploads_async.py:299]: I'm surprised this is working at the moment?? I guess it doesn't matter since it just changes the resultant `block_id`, but the random contributor did not give a properly equivalent statement to the original format string: `block_id = 'BlockId{}'.format("%05d" % (index/self.chunk_size))`

Sample Code:
```index = 10
chunk_size = 5

block_id = 'BlockId{}'.format("%05d" % (index/chunk_size))
block_id2 = f'BlockId{index/chunk_size:05%}'
block_id3 = f'BlockId{"%05d" % (index/chunk_size)}'

print('Original format:', block_id)
print("F-string from random contributor:", block_id2)
print("Fixed f-string:", block_id3)
```

Sample Output:
```
Original format: BlockId00002
F-string from random contributor: BlockId200.000000%
Fixed f-string: BlockId00002
```

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/base_client_async.py [hunk 9]
```diff
@@ -70,13 +69,13 @@ def _create_pipeline(self, credential, **kwargs):
         # type: (Any, **Any) -> Tuple[Configuration, Pipeline]
         self._credential_policy = None
         if hasattr(credential, 'get_token'):
-            self._credential_policy = AsyncBearerTokenCredentialPolicy(credential, STORAGE_OAUTH_SCOPE)
+            self._credential_policy = AsyncStorageBearerTokenCredentialPolicy(credential)
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/base_client_async.py:72]: As discussed, let's not take this specific change to avoid introducing Bearer challenges to other packages. Looks like you may have just missed reverting this for async?

Applies to file-share and queue as well.

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/uploads.py (hunk 3 above)
Comment (jalauzon-msft) [reply sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/uploads.py:601]: Cool, yeah looks like a change that should come to Datalake.

File: sdk/storage/azure-storage-queue/azure/storage/queue/_shared/request_handlers.py (hunk 7 above)
Comment (jalauzon-msft) [reply sdk/storage/azure-storage-queue/azure/storage/queue/_shared/request_handlers.py:74]: Yep, should be fine. Looks like a customer made this change and only added to blob.

File: sdk/storage/azure-storage-queue/azure/storage/queue/_shared/uploads_async.py (hunk 8 above)
Comment (jalauzon-msft) [reply sdk/storage/azure-storage-queue/azure/storage/queue/_shared/uploads_async.py:299]: Wowza, good catch.

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/response_handlers.py (hunk 1 above)
Comment (vincenttran-msft) [reply sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/response_handlers.py:69]: Take more robust one from `deserialize.py` and slap it into `_shared` (this change needs to be done in dlake & queue)

**Review by jalauzon-msft — APPROVED** [review]
Looks good, thanks for working through this!
